data/jsonic1_commit.py

```diff
@@ -0,0 +1,70 @@
+Grabs data from the "FAQ Content" CSV and turns it into nice JSON: a main faq object containing an array of themed Section objects, each Section Object in turn holding an array of Question objects consisting of question, answer and related link/s, as follows:
+    [
+        {
+          "Q": "What are the symptoms of COVID-19?",
+          "A": "Symptoms of COVID-19 include coughing and shortness of breath. Additionally, a person showing two or more of the following symptoms may have the virus: fever, repeated shaking with chills, muscle pain, headache, sore throat, new loss of taste or smell.",
+          "link": "https://www.cdc.gov/coronavirus/2019-ncov/symptoms-testing/symptoms.html"
+        },
+        
+    .
+    .
+    .
+    ]
+    etc
+In addition to its array of Question objects, each Section displays a title and "last updated" date value. For the Python stage of the transition from CSV to JSON we create a hierarchy of nested dictionaries and lists equivalent to the JSON objects and arrays.
+"""
+
+import requests
+import json
+import csv
+from io import StringIO
+import datetime
+
+# Link to "FAQ Content" CSV
+link = "https://docs.google.com/spreadsheets/d/1_wBXS62S5oBQrwetGc8_-dFvDjEmNqzqHwUeP-DzkYs/export?format=csv&id=1_wBXS62S5oBQrwetGc8_-dFvDjEmNqzqHwUeP-DzkYs&gid=1318925039"
+
+# Get the data
+response = requests.get(link)
+response.raise_for_status()
+
+# Turn text from CSV stream into a file object csv.reader can read
+csv_stream = StringIO(response.text)
+# Create a reader to iterate over the CSV stream file object
+reader = csv.reader(csv_stream)
+
+# ISO8601 format date, to update each time the program runs
+date = datetime.datetime.now().strftime("%Y-%m-%d")
+
+# Create the main FAQ dictionary
+# Create a list for our Sections
+# Place the Sections list inside the main dictionary
+faq_dict = {}
+sections_list = []
+faq_dict["faqItems"] = sections_list
+
+# Get data from the reader object
+# Arrange Sections and Questions in a JSON-friendly way
+for row in reader:
+    rowtype = row[0]   # CSV column delineating Section, Question, Answer or Link
+    rowval = row[1]   # Values for the above
+    if rowtype == "Section Head":
+        section = {}   # Create a dictionary for a Section      
+        sections_list.append(section)   # Add new Section to the Sections list
+        section["title"] = rowval   # Give the Section its title
+        section["lastUpdateAt"] = date   # a "last updated" value
+        questions_list = []   # Create a list for Questions
+        section["qa"] = questions_list   # add our Questions list to the Section
+    if rowtype == "Q":        
+        question = {}   # Create new Question object
+        question["Q"] = rowval   # the Question's title
+        questions_list.append(question)   #append the Question to its given Section
+    elif rowtype == "A":   # Now the same for Answers & Links
+        question["A"] = rowval
+    elif rowtype == "link":
+        question["link"] = rowval
```

Comment:
It would also be a good idea to add an `else` clause to stop this script and warn you if you hit an unexpected value. Using `enumerate` for the for loop can also help provide nicer messaging:

```py
for index, row in enumerate(reader):
    # The code you already have here, then...
    elif rowtype == '':
        pass  # Empty rows are OK
    else:
        raise ValueError(f'Unknown header for row {index}: "{rowtype}"')
```

Reply:
Oh, I also just noticed some lines have no text in the header column, but do have values. e.g. lines 51-53. Is that an error? If so, maybe have a branch like:

```py
    elif rowvalue and not rowtype:
        raise ValueError(f'Row {index} has a value, but no header to describe it')
```

Reply:
Rob - thanks so much for this: I've learnt a bunch of new tricks and approaches here (exciting!). So much tighter and neater :-)
Agree re the data directory (and requirements)...this wants to be where the code folk can easily find it I guess.

Reply:
One more query: would this require reverting to the original code for indexing the columns, or would the new for loop contain the for more svelte for loop suggested above ( ie 'for rowtype, rowval, *_ in reader:'). If so, would the extra elifs and else sit outside the original for loop? And do those questions make any sense whatsoever? Is the earth actually flat, or are they just making that up? And where do the odd socks go? You can see how difficult this is to understand...

Reply:
So the extra `elif`s and `else` I’m suggesting would be part of the main `if` statement you already have, i.e. inside the `for` loop. So you had:

```py
for row in reader:
    rowtype, rowval, *_ = row
    if rowtype == "Section Head":
        ...
    if rowtype == "Q":
        ...
    elif rowtype == "A":
        ...
    elif rowtype == "link":
        ...
```

And I was suggesting:

```py
for row in reader:
    rowtype, rowval, *_ = row
    if rowtype == "Section Head":
        ...
    if rowtype == "Q":
        ...
    elif rowtype == "A":
        ...
    elif rowtype == "link":
        ...
    elif rowvalue and not rowtype:
        raise ValueError('Row has a value, but no header to describe it')
    elif rowtype == '':
        pass  # Empty rows are OK
    else:
        raise ValueError(f'Unknown row header: "{rowtype}"')
```

As far as more svelte `for` loop suggested above (i.e. `for rowtype, rowval, *_ in reader:`), you can combine that with `enumerate` like so:

```py
for index, [rowtype, rowval, *_] in enumerate(rows):
    # Your massive if statment here
```

Or you can keep it separate:

```py
for index, row in enumerate(rows):
    rowtype, rowval, *_ = row
    # Your massive if statment here
```

Reply:
Rob - thanks SO much for taking the time to explain these things so well (much more clearly than much of the documentation I was consulting yesterday in fact!). The Python docs for the enumerate built-in actually point out that even a lot of experienced Pythonistas overlook it, so this is a lovely thing to learn about - as with the other refinements you've introduced. Gunna spend time tonight reading and practicing, then incorporate it all into the script and create a fresh PR in the data channel for double-checking. I def owe you a beer when such a thing is possible again :-)